### DISCUSSION TITLE
A BUG for DataGridCheckBoxColumn of Avalonia.Controls.DataGrid

### DISCUSSION THREAD

**Discussion by jioyou** [AvaloniaUI/Avalonia#14715] [Q&A]
## Expected behavior
I'm running into a small bug when using Avalonia.Controls.DataGrid's DataGridCheckBoxColumn. 
After clicking the checkbox of the DataGridCheckBoxColumn column, 
click again to unselect it, 
but there is no response. 
You need to click on another cell and then click checkbox to unselect it.

## Screenshots
![456](https://github.com/AvaloniaUI/Avalonia/assets/13229266/d07d4157-1725-4d8b-ad44-ac660541a5f0)

## Environment 

- OS: Windows
- Avalonia-Version: [e.g. 11.0.7 and 11.0.9]

## Additional context
11.0.5 and earlier versions is fine, 11.0.7 and later versions have this issue

**Comment by timunie** [comment]
Could you check nightly builds please?

**Comment by timunie** [comment]
I cannot reproduce it using 11.0.9. Please file a minimal repro. Thx.

**Reply by timunie** [reply]
made it a discussuion for now as it is not reproducible for me

**Reply by khanhstitsel** [reply]
Hello, 

I have this exact error, binding to a very simple collection, here are some of the codes:
Axaml:
<DataGrid ItemsSource="{Binding ExtraPoints}" Grid.Row="0" Grid.Column="2">

	<DataGrid.Columns>
		<DataGridCheckBoxColumn Header="Check to add" Binding="{Binding Included}"
								Width="*"/>
		<DataGridTextColumn Header="Tp Number" Binding="{Binding TestPointNumber}"
							Width="*" Foreground="Black"/>
	</DataGrid.Columns>
</DataGrid>

ViewModel:

	public class ExtraPointsViewModel : ViewModelBase{
		public ExtraPointsViewModel() {
			ExtraPoints = new() {
				new(true, 1),
				new(true, 2),
			};
		}
		public ObservableCollection<ExtraPoint> ExtraPoints { get; }
	}
	
Model:

	public class ExtraPoint {
		public ExtraPoint(Boolean included, Int32 testPointNumber) {
			Included = included;
			TestPointNumber = testPointNumber;
		}

		public ExtraPoint() { }
		public Boolean Included { get; set; } = false;
		public Int32 TestPointNumber { get; set; } = 0;
	}

Everytime I check a box or uncheck it, I need to click another element before the same box can be check/uncheck again. It looks like [this issue](https://github.com/AvaloniaUI/Avalonia/issues/10233) that was resolved previously, but for some reason I encounter it on my first day with Avalonia. The version that is specified in my Visual Studio Nuget Package Manager is 11.0.10. Thank you.

**Reply by timunie** [reply]
ExtraPoint needs to raise PropertyChanged

**Reply by khanhstitsel** [reply]
I implemented your suggestions just in case, but even before that, it is not really a matter of binding that I am encountering. Checkboxes that are inside of data grid columns behaves totally different from the checkboxes that are outside. 

After the changes were made, the UI behaviors are still the same. Once checked or unchecked, even if I double click the same data grid column checkbox again multiple time, it will not go to the opposite state; I need to click on another window, or click on some other element within the same form - clicking on white space within the same form does not resolve the inability of checkbox to switch state. Normal checkbox component works fine as expected, I can keep clicking on the same checkbox to repeatedly switch state.

The code after your suggestion is here just in case, there is no functions that changes the states of Included in the code behind:

    public class ExtraPoint : INotifyPropertyChanged{
	public ExtraPoint(Boolean included, Int32 testPointNumber) {
		Included = included;
		TestPointNumber = testPointNumber;
	}

	public event PropertyChangedEventHandler PropertyChanged;

	protected virtual void OnPropertyChanged(string propertyName) {
		PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
	}

	public ExtraPoint() { }

	Boolean _included = false;
	public Boolean Included {
		get => _included;
		set {
			_included = value;
			OnPropertyChanged(nameof(Included));
		} 
	}

	Int32 _testPointNumber = 0;
	/// <summary>Gets or sets the test point number.</summary>
	public Int32 TestPointNumber {
		get => _testPointNumber;
		set {
			_testPointNumber = value;
			OnPropertyChanged(nameof(TestPointNumber));
		}
	}
}

**Reply by timunie** [reply]
ah okay that's the confusion. So the DataGrid is specail here as it needs a double-click to make the cell entering edit mode and a lost focus to exit edit mode. Then the binding gets updated. If you dislike that, I suggest to use a `DataGridTemplateColumn` where you can add an editable CheckBox directly.

**Reply by khanhstitsel** [reply]
I see, so that is intended behavior? Not sure if I ever use any GUI that have such a strange checkbox behavior. Anyhow, thank you very much for your time.

**Reply by timunie** [reply]
DataGrid was isnpired by WPF and Silverlight afaik. 

- https://stackoverflow.com/questions/10651206/datagridcheckboxcolumn-two-way-binding

in 11.1 we will also have UpdateSourceTrigger like WPF has but we haven't it released yet.
